File: common/cpp/gs/net/upstream.cpp
#include "upstream.hpp"
#include <iostream>

namespace gs {
namespace net {

UpstreamPool::UpstreamPool(DataCallback on_data) : on_data_(std::move(on_data)) {}

UpstreamPool::~UpstreamPool() {
    Stop();
}

void UpstreamPool::AddNode(const std::string& host, uint16_t port) {
    std::lock_guard<std::mutex> lk(nodes_mtx_);
    auto node = std::make_unique<UpstreamNode>();
    node->host = host;
    node->port = port;
    node->client = std::make_unique<TCPClient>(host, port);
    nodes_.push_back(std::move(node));
}

bool UpstreamPool::Start() {
    if (nodes_.empty()) {
        std::cerr << "UpstreamPool: no nodes configured" << std::endl;
        return false;
    }

    // 尝试初始连接所有节点
    {
        std::lock_guard<std::mutex> lk(nodes_mtx_);
        for (auto& node : nodes_) {
            TryConnect(node.get());
        }
    }

    running_ = true;
    reconnect_thread_ = std::thread([this] { ReconnectLoop(); });
    return true;
}

void UpstreamPool::Stop() {
    running_ = false;
    reconnect_cv_.notify_all();
    if (reconnect_thread_.joinable()) {
        reconnect_thread_.join();
    }

    std::lock_guard<std::mutex> lk(nodes_mtx_);
    for (auto& node : nodes_) {
        if (node->client) {
            node->client->Close();
        }
    }
}

UpstreamNode* UpstreamPool::Pick() {
    std::lock_guard<std::mutex> lk(nodes_mtx_);
    if (nodes_.empty()) return nullptr;

    // 轮询：从当前索引开始找第一个健康的
    size_t start = rr_index_.fetch_add(1) % nodes_.size();
    for (size_t i = 0; i < nodes_.size(); ++i) {
        size_t idx = (start + i) % nodes_.size();
        if (nodes_[idx]->healthy.load() && nodes_[idx]->client && nodes_[idx]->client->IsConnected()) {
            nodes_[idx]->last_active = std::chrono::steady_clock::now();
            return nodes_[idx].get();
        }
    }
    return nullptr;
}

bool UpstreamPool::SendPacket(const Packet& pkt) {
    auto* node = Pick();
    if (!node || !node->client || !node->client->Conn()) {
        return false;
    }
    return node->client->Conn()->SendPacket(pkt);
}

void UpstreamPool::BroadcastToHealthy(const Packet& pkt) {
    std::lock_guard<std::mutex> lk(nodes_mtx_);
    for (auto& node : nodes_) {
        if (node->healthy.load() && node->client && node->client->Conn()) {
            node->client->Conn()->SendPacket(pkt);
        }
    }
}

size_t UpstreamPool::HealthyCount() const {
    std::lock_guard<std::mutex> lk(nodes_mtx_);
    size_t count = 0;
    for (const auto& node : nodes_) {
        if (node->healthy.load()) ++count;
    }
    return count;
}

size_t UpstreamPool::TotalCount() const {
    std::lock_guard<std::mutex> lk(nodes_mtx_);
    return nodes_.size();
}

void UpstreamPool::ReconnectLoop() {
    int interval_ms = 1000; // 初始 1s，指数退避到 30s
    const int max_interval_ms = 30000;

    while (running_.load()) {
        std::unique_lock<std::mutex> lk(reconnect_mtx_);
        reconnect_cv_.wait_for(lk, std::chrono::milliseconds(interval_ms),
                               [this] { return !running_.load(); });
        if (!running_.load()) break;

        bool any_reconnected = false;
        {
            std::lock_guard<std::mutex> node_lk(nodes_mtx_);
            for (auto& node : nodes_) {
                if (!node->healthy.load() && !node->connecting.load()) {
                    if (TryConnect(node.get())) {
                        any_reconnected = true;
                    }
                }
            }
        }

        if (any_reconnected) {
            interval_ms = 1000;
        } else if (HealthyCount() == 0) {
            interval_ms = std::min(interval_ms * 2, max_interval_ms);
        } else {
            interval_ms = 1000;
        }
    }
}

bool UpstreamPool::TryConnect(UpstreamNode* node) {
    node->connecting.store(true);
    bool ok = node->client->Connect();
    if (ok) {
        node->healthy.store(true);
        node->last_active = std::chrono::steady_clock::now();
        std::cout << "UpstreamPool: connected to " << node->host << ":" << node->port << std::endl;

        // 设置关闭回调：节点断开时标记为不健康
        node->client->SetCallbacks(
            on_data_,
            [node](TCPConn*) {
                node->healthy.store(false);
                std::cerr << "UpstreamPool: disconnected from " << node->host << ":" << node->port << std::endl;
            }
        );
    } else {
        node->healthy.store(false);
    }
    node->connecting.store(false);
    return ok;
}

} // namespace net
} // namespace gs

File: common/go/net/upstream.go
package net

import (
	"sync"
	"sync/atomic"
	"time"
)

// Node 上游节点运行时状态
type Node struct {
	Addr      string
	Client    *TCPClient
	Healthy   atomic.Bool
	Connecting atomic.Bool
	LastActive time.Time
}

// UpstreamPool 上游连接池：多节点、轮询、自动重连
type UpstreamPool struct {
	mu       sync.RWMutex
	nodes    []*Node
	rrIndex  atomic.Uint64

	onData   func(*TCPConn, *Packet)
	onNodeEvent func(addr string, healthy bool) // 节点健康状态变更回调

	running  atomic.Bool
	stopCh   chan struct{}
	wg       sync.WaitGroup
}

// NewUpstreamPool 创建连接池
func NewUpstreamPool(onData func(*TCPConn, *Packet)) *UpstreamPool {
	return &UpstreamPool{
		onData: onData,
		stopCh: make(chan struct{}),
	}
}

// SetOnNodeEvent 设置节点健康状态变更回调
func (p *UpstreamPool) SetOnNodeEvent(fn func(addr string, healthy bool)) {
	p.onNodeEvent = fn
}

// AddNode 添加上游节点（Start 前调用）
func (p *UpstreamPool) AddNode(addr string) {
	p.mu.Lock()
	defer p.mu.Unlock()
	for _, n := range p.nodes {
		if n.Addr == addr {
			return // 已存在
		}
	}
	n := &Node{Addr: addr}
	n.Healthy.Store(false)
	n.Connecting.Store(false)
	p.nodes = append(p.nodes, n)
}

// SetOnData 动态设置收到数据时的回调（用于 Registry 发现模式后绑定 rpc.Client）
func (p *UpstreamPool) SetOnData(onData func(*TCPConn, *Packet)) {
	p.mu.Lock()
	p.onData = onData
	p.mu.Unlock()
}

// RemoveNode 移除节点
func (p *UpstreamPool) RemoveNode(addr string) {
	p.mu.Lock()
	defer p.mu.Unlock()
	for i, n := range p.nodes {
		if n.Addr == addr {
			if n.Client != nil {
				n.Client.Close()
			}
			p.nodes = append(p.nodes[:i], p.nodes[i+1:]...)
			return
		}
	}
}

// Start 启动连接池（尝试初始连接 + 后台重连）
func (p *UpstreamPool) Start() {
	p.mu.Lock()
	for _, n := range p.nodes {
		p.tryConnectLocked(n)
	}
	p.mu.Unlock()

	p.running.Store(true)
	p.wg.Add(1)
	go p.reconnectLoop()
}

// Stop 停止连接池
func (p *UpstreamPool) Stop() {
	if !p.running.CompareAndSwap(true, false) {
		return
	}
	close(p.stopCh)
	p.wg.Wait()

	p.mu.Lock()
	defer p.mu.Unlock()
	for _, n := range p.nodes {
		if n.Client != nil {
			n.Client.Close()
		}
	}
}

// Pick 轮询选择一个健康节点，返回 nil 表示无可用节点
func (p *UpstreamPool) Pick() *Node {
	p.mu.RLock()
	defer p.mu.RUnlock()
	if len(p.nodes) == 0 {
		return nil
	}

	start := int(p.rrIndex.Add(1)) % len(p.nodes)
	for i := 0; i < len(p.nodes); i++ {
		idx := (start + i) % len(p.nodes)
		n := p.nodes[idx]
		if n.Healthy.Load() && n.Client != nil && n.Client.Conn() != nil {
			n.LastActive = time.Now()
			return n
		}
	}
	return nil
}

// SendPacket 发送 Packet 到选中的健康节点
func (p *UpstreamPool) SendPacket(pkt *Packet) bool {
	n := p.Pick()
	if n == nil || n.Client == nil || n.Client.Conn() == nil {
		return false
	}
	return n.Client.Conn().SendPacket(pkt)
}

// HealthyCount 当前健康节点数
func (p *UpstreamPool) HealthyCount() int {
	p.mu.RLock()
	defer p.mu.RUnlock()
	count := 0
	for _, n := range p.nodes {
		if n.Healthy.Load() {
			count++
		}
	}
	return count
}

// TotalCount 总节点数
func (p *UpstreamPool) TotalCount() int {
	p.mu.RLock()
	defer p.mu.RUnlock()
	return len(p.nodes)
}

// AllNodes 返回所有节点地址（用于外部遍历）
func (p *UpstreamPool) AllNodes() []*Node {
	p.mu.RLock()
	defer p.mu.RUnlock()
	out := make([]*Node, len(p.nodes))
	copy(out, p.nodes)
	return out
}

func (p *UpstreamPool) reconnectLoop() {
	defer p.wg.Done()
	interval := time.Second
	maxInterval := 30 * time.Second

	for {
		select {
		case <-p.stopCh:
			return
		case <-time.After(interval):
		}

		anyReconnected := false
		p.mu.Lock()
		for _, n := range p.nodes {
			if !n.Healthy.Load() && !n.Connecting.Load() {
				if p.tryConnectLocked(n) {
					anyReconnected = true
				}
			}
		}
		p.mu.Unlock()

		if anyReconnected {
			interval = time.Second
		} else if p.HealthyCount() == 0 {
			interval = min(interval*2, maxInterval)
		} else {
			interval = time.Second
		}
	}
}

func (p *UpstreamPool) tryConnectLocked(n *Node) bool {
	n.Connecting.Store(true)
	client := NewTCPClient(n.Addr, p.onData, func(_ *TCPConn) {
		n.Healthy.Store(false)
		if p.onNodeEvent != nil {
			p.onNodeEvent(n.Addr, false)
		}
	})
	if err := client.Connect(); err != nil {
		n.Healthy.Store(false)
		n.Connecting.Store(false)
		return false
	}
	n.Client = client
	n.Healthy.Store(true)
	n.LastActive = time.Now()
	n.Connecting.Store(false)
	if p.onNodeEvent != nil {
		p.onNodeEvent(n.Addr, true)
	}
	return true
}

func min(a, b time.Duration) time.Duration {
	if a < b {
		return a
	}
	return b
}
